ESP/lib/src/io/Improv/improv.cpp

```diff
@@ -0,0 +1,155 @@
+#include "improv.h"
+
+namespace improv {
+
+  ImprovCommand parse_improv_data(const std::vector<uint8_t>& data,
+                                  bool check_checksum) {
+    return parse_improv_data(data.data(), data.size(), check_checksum);
+  }
+
+  ImprovCommand parse_improv_data(const uint8_t* data,
+                                  size_t length,
+                                  bool check_checksum) {
+    ImprovCommand improv_command;
+    Command command = (Command)data[0];
+    uint8_t data_length = data[1];
+
+    if (data_length != length - 2 - check_checksum) {
+      improv_command.command = UNKNOWN;
+      return improv_command;
+    }
+
+    if (check_checksum) {
+      uint8_t checksum = data[length - 1];
+
+      uint32_t calculated_checksum = 0;
+      for (uint8_t i = 0; i < length - 1; i++) {
+        calculated_checksum += data[i];
+      }
+
+      if ((uint8_t)calculated_checksum != checksum) {
+        improv_command.command = BAD_CHECKSUM;
+        return improv_command;
+      }
+    }
+
+    if (command == WIFI_SETTINGS) {
+      uint8_t ssid_length = data[2];
+      uint8_t ssid_start = 3;
+      size_t ssid_end = ssid_start + ssid_length;
+
+      uint8_t pass_length = data[ssid_end];
+      size_t pass_start = ssid_end + 1;
+      size_t pass_end = pass_start + pass_length;
+
+      std::string ssid(data + ssid_start, data + ssid_end);
+      std::string password(data + pass_start, data + pass_end);
+      return {.command = command, .ssid = ssid, .password = password};
+    }
+
+    improv_command.command = command;
+    return improv_command;
+  }
+
+  bool parse_improv_serial_byte(size_t position,
+                                uint8_t byte,
+                                const uint8_t* buffer,
+                                std::function<bool(ImprovCommand)>&& callback,
+                                std::function<void(Error)>&& on_error) {
+    if (position == 0)
```

Comment:
to ja bym zwinął do jakiegoś
`if (position <=5)`
a tekst `"IMPROV"` mieć w jakiejś stałej `text` i porównywać `text[position] == byte`

W przyszłości łatwiej będzie znaleźć "improv" niż pojedyncze literki

Reply:
Indeed,  i think we should move changes to separate branch. 

The `improv.cpp` and `improv.h` were not written by me.  They were taken directly out of there github repo - as their docs state. I _definitely_ wouldn't have written it like that. There are quite a few things i would change. I didn't want to change anything before i got it working.

Reply:
I only wrote the `improvHandler` files.

Reply:
Yeah, I figured something either went really wrong, ooor those weren't your changes, in that case that's a bit concerning that they did it this way, but oh well